### PR TITLE
parser:Support for setting global variable with `on` and `off`

### PR DESCRIPTION
**Pull request by louishust**

### CODE REVIEW

**Comment by louishust** [comment]
related to #3581

File: executor/set_test.go [hunk 1]
```diff
@@ -123,6 +123,12 @@ func (s *testSuite) TestSetVar(c *C) {
 	c.Assert(vars.SkipConstraintCheck, IsTrue)
 	tk.MustExec("set @@tidb_skip_constraint_check = '0'")
 	c.Assert(vars.SkipConstraintCheck, IsFalse)
+
+	tk.MustExec("set global avoid_temporal_upgrade = on")
```
Comment (shenli) [inline executor/set_test.go]: Any test case about "off"?

File: executor/set_test.go [hunk 2]
```diff
@@ -123,6 +123,12 @@ func (s *testSuite) TestSetVar(c *C) {
...
+	tk.MustExec("set session sql_log_bin = on")
+	tk.MustExec("set sql_log_bin = on")
+	tk.MustExec("set @@sql_log_bin = on")
+	tk.MustExec("set @@global.sql_log_bin = on")
```
Comment (shenli) [inline executor/set_test.go]: Check the result of those cases. Such as 'select @@global.sql_log_bin'.

**Comment by louishust** [comment]
@shenli PTAL

File: executor/set_test.go (hunk 1 above)
Comment (breeswish) [inline executor/set_test.go]: why do `set xxx on` becomes an upper cased `ON` while `set xxx off` becomes a lower cased `off`?

Comment (louishust) [inline executor/set_test.go]: cause ON is keyword in SQL syntax, and off is not keyword, ON is handled in parser.y specially.

Comment (breeswish) [inline executor/set_test.go]: Ok, I opened another issue for this: https://github.com/pingcap/tidb/issues/3674

Comment (louishust) [inline executor/set_test.go]: yes, how to store variables should be in another issue. For now , tidb just store the value in table without really parse and handle these variables.

Comment (zz-jason) [inline executor/set_test.go]: how about using "true" or "false" ?

Comment (louishust) [inline executor/set_test.go]: true or false have been handled in parser.y in Literal.
This issue only handle the syntax error produced by SET variable = ON

**Comment by breeswish** [comment]
LGTM

**Comment by tiancaiamao** [comment]
> cause ON is keyword in SQL syntax, and off is not keyword, ON is handled in parser.y specially.

```
set global avoid_temporal_upgrade = on;
```

and 

```
set global avoid_temporal_upgrade = 'on';
```

They are not the same thing. `on` is a special token, while `'on'` is a stringLit.

When you write

```
set global avoid_temporal_upgrade = 'on';
```

`'on'` is a turn to `Literal` -> `Operand` -> `PrimaryExpression` -> `PrimaryFactor` -> `PredicateExpr` -> `Factor` -> `Expression`  and finally it become part of `VariableAssignment`.


When you write

```
set global avoid_temporal_upgrade = on;
```

There is no rule for it before this PR, so it failed.

I believe 

```
set global avoid_temporal_upgrade = off;
```

is not handled at all in this PR.

**Comment by breeswish** [comment]
@tiancaiamao 

As I tested previously, `set global avoid_temporal_upgrade = off;` is working even without this PR: `off` becomes `"off"`. In fact you can write `set global avoid_temporal_upgrade = hahaha` and it becomes `set global avoid_temporal_upgrade = "hahaha"`.....

**Comment by louishust** [comment]
@tiancaiamao @breeswish 
This PR just resolve the syntax error of  `set global xxx = ON `.
Cause ON is keyword and OFF is just handled as identifier but not keyword.

**Comment by tiancaiamao** [comment]
```
set global xxx = off
```

`'off'` -> `identifier` -> `Identifier` -> `ColumnName` -> `Operand` -> `PrimaryExpression` -> `PrimaryFactor` -> `PredicateExpr` -> `Factor` -> `Expression` -> `VariableAssignment`

This is a bug, we should fix it.

**Comment by breeswish** [comment]
However the behavior of "auto converting an unknown token to a string value" should be preserved because it is supported by MySQL : /

**Comment by tiancaiamao** [comment]
> "auto converting an unknown token to a string value" 

Where is the references? @breeswish

**Comment by breeswish** [comment]
@tiancaiamao

```
mysql> set global init_connect = helloworld;
Query OK, 0 rows affected (0.00 sec)

mysql> select @@global.init_connect;
+-----------------------+
| @@global.init_connect |
+-----------------------+
| helloworld            |
+-----------------------+
1 row in set (0.00 sec)
```

**Comment by louishust** [comment]
For now, tidb just store all variables in string format,which is very different with MySQL.
MySQL store sql_log_bin as bool variables,and store init_connect as string variables.
So that if you `set sql_log_bin= helloworld` will lead to error.
This PR just want to resolve the syntax error for set global xxx =ON, and store ON as the value.
And if we want to handle variable values like what mysql does, it should be a big change in tidb for all variables.

**Comment by breeswish** [comment]
I would like to mention that, when executing `set sql_log_bin = helloworld`, you will receive an error which is not raised from the parser but from the executor, since the error message is `ERROR 1231 (42000): Variable 'sql_log_bin' can't be set to the value of 'helloworld'` (as you can see, it has already converted the token `helloworld` into a string value `'helloworld'`).

**Comment by louishust** [comment]
tidb parse filed `helloworld` as string item is the same as how mysql handles.

**Comment by breeswish** [comment]
@louishust Yes, that's why I think it would be better to keep this behavior.

**Comment by louishust** [comment]
@breeswish TIDB can keep the behavior, and what tidb should to do is to store different variables in different format, such as bool ,string, integer and do check for each variables like what mysql does.

**Comment by breeswish** [comment]
@louishust Agree. https://github.com/pingcap/tidb/issues/3674

**Comment by tiancaiamao** [comment]
> I think it would be better to keep this behavior.

```
set sql_log_bin = hello world;
```

I agree that it should parse without error.

But...the generated ast should not be `&ast.ColumnNameExpr{Name: $1.(*ast.ColumnName)}`, this is bug, without doubt.

Would you like to fix it? @louishust @breeswish 


> in summary, at the parser level MySQL will auto convert those tokens into string values.

This is not a considered judgments.
No document mentioned it. The test is done in a special context, without even know why.
If I can't understand why an `identifier` promote to an `expression`, I can't prove it's a correct change.

**Comment by louishust** [comment]
@tiancaiamao So tidb should convert ColumnNameExpr to NewValueExpr in set stmt?

**Comment by louishust** [comment]
@tiancaiamao 

MySQL convert FIELD_ITEM to String_ITEM when FIELD_ITEM in the `value` part of set statement.

```
set_var::set_var(enum_var_type type_arg, sys_var *var_arg,
                 const LEX_STRING *base_name_arg, Item *value_arg)
  :var(var_arg), type(type_arg), base(*base_name_arg)
{
  /*
    If the set value is a field, change it to a string to allow things like
    SET table_type=MYISAM;
  */
  if (value_arg && value_arg->type() == Item::FIELD_ITEM)
  {
    Item_field *item= (Item_field*) value_arg;
    if (item->field_name)
    {
      if (!(value= new Item_string(item->field_name,
                                   strlen(item->field_name),
                                   system_charset_info))) // names are utf8
        value= value_arg;			/* Give error message later */
    }
    else
```

You can check the code at sql/set_var.cc of mysql5.7 source code

**Comment by zimulala** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

This PR get the right result but not the best way of the implementation, due to some historical reasons. I would refactor `parser.y` and fix some unreasonable part when I have time.

**Pull request merged**